Edit plugin

sampsyo opened the pull request:
Successor to #1687. This is @jmwatte's interactive editor plugin.

sampsyo commented:
@jmwatte, I'm interested in trimming down the initial version of this plugin to the bare minimum. We can add more flexibility later, but I'm really interested in getting something as simple as possible for the first iteration.

Would it be OK with you if I removed some of the indirection that supports multiple methods? For now, I'll hard-wire this to be YAML-only for readability's sake.

sampsyo commented:
I'm also not 100% sure about including your (rather clever) logic for reducing the display of fields with the same value (the `sum` option). It's clearly going to be useful eventually, but it requires a lot of logic that makes the first version a bit harder to read. Would it also be OK if I temporarily removed that for the sake of the initial merge?

sampsyo commented on beetsplug/edit.py:
```diff
@@ -0,0 +1,438 @@
+            callmethod.append(name)
+            subprocess.check_call(callmethod)
+
+    def same_format(self, newset, opts):
```
@jmwatte, can you explain a little more (either here or in comments) what `same_format` and `save_items` are doing? The short variable names and lack of comments/docstrings is making it difficult for me to follow the logic here.

jmwatte replied:
Sure.
2 different objects could come in the save method. 1 from sum 1 from your normal view. They have different structures so I make em look the same in same_format. I also make it print the title/artist instead of the Id. But if you delete the sum part you will probably not need that.

diego-plan9 commented on beetsplug/edit.py:
```diff
@@ -0,0 +1,438 @@
+            if editor:
+                os.system(editor + " " + name)
+            else:
+                webbrowser.open(name, new=2, autoraise=True)
```
Just a quick question in regards to the safety of using `webbrowser.open` I came across while working on my previous set of changes: has it been reasonably tested that it really produces the desired results? Specially, considering the remark on [the python docs](https://docs.python.org/2/library/webbrowser.html#webbrowser.open):

>  Note that on some platforms, trying to open a filename using this function, may work and start the operating system’s associated program. However, this is neither supported nor portable.

For example, in my home machine (debian 8 + libreoffice without too much tuning), seems that debian by default does not add the `$EDITOR` variable to the environment, and as a consequence `webbrowser.open` launchs libreoffice writer.

If associating .yaml files with vim via `$ xdg-mime default vim.desktop application/x-yaml`, webbrowser seems to exhibit some weird behaviour, launching vim but not replying to stdin correctly and resulting in an unresponsive vim - after killing it manually, it launches libreoffice once again. Might be worth devoting some time to check it more carefully, as it might affect more users and the `config` command as well?

Perhaps it would be enough to add a clearer reminder to the users on the documentation to set `$EDITOR`, to avoid nuisances? It's already on http://beets.readthedocs.org/en/v1.3.15/reference/cli.html#config anyway, but might not hurt.

diego-plan9 commented:
In the hopes of joining efforts with this neat @jmwatte feature, I'm thinking of starting by writing some tests, as it would help me getting up to speed with the code and seems like they will be needed anyway. Would that be a good idea, or is there any more urgent task where I can lend a hand?

Also, what would be the preferred way to push my changes? I have synced my cloned repository but I'm not sure if issuing a new pull request from my `edit` branch to the main repository `edit` branch would be automatically appended to this one or instead create a new one, adding some "noise" in the process.

sampsyo commented:
Yes! Acceptance tests would be a good place to start. Other than this, my main priority right now is simplification.

The best way to contribute for now would be to send pull requests against the `editor` branch. When those are merged, the associated commits should automatically get added to the main PR.

diego-plan9 commented:
Thanks for adding me as a collaborator, @sampsyo! I'll try to commit directly in case of minor changes (such as my latest 6e6aa97 commit), but will probably still use pull requests for more complex changes until I'm more confident in my beets knowledge.

I have hardened the existing tests a bit (which explicit assertions about what fields are changed after the command is run), and included a couple of ones: `test_single_title_edit_apply()`, which tests the modification of a single item by the user even when the yaml contains several items, and `test_album_edit_apply()`, testing the modification of the `album` field.

I'm wondering if the plugin should we expected to be "intelligent" enough at this stage to cope with changes on `item.album`, propagating them to the album accordingly? At this moment it seems corresponding `lib.album` is not changed at all, which might be fine at this stage but seems like a desirable behaviour.

jmwatte commented:
@diego-plan9 
Hey diego,
Concerning the changes to the album level. Many have the been the questions from users that have changed things at item-level and then being advised to change the albumartist at album-level for desired results.
Been there myself a couple of times.
But I think that  - that's what the plugin is about - since we are talking about manually making changes and not letting the machinery doing it for you we're bound to get into Oh-I-screwed-up-country. With great freedom comes great ... If I wanted to assign a couple of songs to another album, because I like to make my own compilations, I would want that freedom. The difference between album-level and item-level ... it's a sneaky one especially for fresh users and even older users. The only real safety I build in is the not being able to change the ID's (and the path) of the items. And maybe there are even cases where I would want to mess with the path. 
Concerning the webbrowser. I saw a method today in beet.utils (open_anything())that could be used for that.
grtz
jm

sampsyo commented:
> I have hardened the existing tests a bit (which explicit assertions about what fields are changed after the command is run), and included a couple of ones: test_single_title_edit_apply(), which tests the modification of a single item by the user even when the yaml contains several items, and test_album_edit_apply(), testing the modification of the album field.

Sounds good!

> I'm wondering if the plugin should we expected to be "intelligent" enough at this stage to cope with changes on item.album, propagating them to the album accordingly? At this moment it seems corresponding lib.album is not changed at all, which might be fine at this stage but seems like a desirable behaviour. (edit: seems that beets edit -a propagates the changes to the items anyway, so might be a matter of recommending the users to use -a if changing album names).

Good question. Now let's say no for now: We'll require an explicit album flag to modify albums. We can revisit this later.

sampsyo commented:
> Concerning the webbrowser. I saw a method today in beet.utils (open_anything())that could be used for that.

Indeed! I believe there is also a function in `util` for finding the user's $EDITOR or equivalent.

diego-plan9 commented:
Thanks to both of you for the answers and the insights, it sounds very reasonable! I'll tweak the existing test and add a couple of ones for testing the output when the query returns albums (`beet edit -a foo`), for sake of completion.

@sampsyo, could you let me know which current configuration options/command line switchs are likely to survive the refactoring and be included in the first release? I'm thinking of adding some tests for the field-related options (`albumfields`, `itemfields`, `--extra`, `--store`, ...).

Also, while waiting for the initial refactor, I'm tempted to include a function for performing some basic validation on the edited yaml (making sure that it has the same fields as the original yaml, same amount of items, etc), geared towards passing the `test_invalid_yaml_discard` test by raising informative exceptions.

diego-plan9 commented on beetsplug/edit.py:
```diff
@@ -0,0 +1,443 @@
+        if not ui.input_yn(ui.colorize('action_default', "Edit? (y/n)"), True):
+            return
+
+        dict_from_objs = self.make_dict[self.pick](self.fields, objs, opts)
```
@jmwatte: I'm curious about the structure chosen to model `dict_from_objs` (returned by `get_all_fields`, `get_selected_fields`). Currently, in not `--sum` mode it looks like a "list of lists of single-key-dicts", ie:

```python
[[{'id': 1}, {'track': 2}, {'title': u'Title'}, {'artist': u'Artist'}, {u'album': u'Album'}], ...]
```
as opposed to the arguably more "natural" "list of dicts" approach (which seems to ultimately be used during `save_items`, after the `same_format` calls):
```python
[{'id': 1, 'track': 2, 'title': u'Title', 'artist': u'Artist', u'album': u'Album'}, ...]
```

Could you let me know if it provides significant readability, performance or any other advantages for the intermediate stages (diffing, reducing, etc), or if the decision was mainly oriented towards the yaml generation (keeping order of fields, formatting)? I'm still trying to fully understand some parts of the plugin, and would really appreciate some insight!

jmwatte replied:
Diego,
The decision-proces has already disappeared in the fog of wine and cigars. I think mainly if it worked it was alright. I think at a certain moment II looked at the output of the Yaml and then used that as a guidance but if you get it running with the more natural list of dict, why not. Might save the extra call to `same_format`. Remember I'm not a programmer, more of a tinker. But ask away, maybe I can shed some light.

diego-plan9 replied:
Thanks a lot for finding the time to reply and for commenting on the issue, @jmwatte. I'm also a victim of the fog more often than not, and any comments are much appreciated in the hopes of shaping up the plugin!

sampsyo commented:
OK, I've done some major refactoring to simplify the plugin to its bare essentials. I've also added a lot of comments and docstrings, removed some implicit data flow through state updates, reduced the "chattiness" of the interface, and changed the YAML format to a single big dictionary instead of lots of little single-key dictionaries.

I'd love to hear feedback from you two (or anyone who's interested)—did I cut too much? Does this work how you expect it to?

The next step before merging will be to go over the documentation page one more time.

jmwatte commented:
Yup. Nailed it. Just text editor instead of test editor (1docstring)(not at my machine right now). This is the most basic thing. Nice.

diego-plan9 commented:
Quite a trimming! It looks good and clean to me, I'm fully on board with keeping the initial version to a bare minimum and then slowly discussing (and hopefully re-introducing) the changes implemented on the original code by @jmwatte and others discussed on the initial ticket. Thanks for the effort on refactoring, it really does neat! I'll update the tests shortly, to keep them up to date.

diego-plan9 commented:
Revised the tests, and added an extra one for testing the addition of extra fields by the user during the yaml editing. Some thoughts, as a couple of tests are failing/resulting in an error:

* `test_single_title_edit_apply` is failing due to the assertion on the number of calls to `Item.write()` (it expects 1 call to `write()`, and recording 10). It seems that if a change is detected in the yaml (even if only a single item is affected), **all** items have their `.write()` method called during `EditPlugin.save_write()`.
It does not look too harmful, but I'm wondering if this behaviour is expected?

* The new test (`test_invalid_yaml_extra_field`, which tries to mimic the user adding random fields to the yaml) results in an error, as the plugin marks it as a valid change during `EditPlugin.save_write()` and tries to act accordingly:
```
the artist - älbum - tïtle 0
  foo: bar
Apply changes? (y/n) 
```
Perhaps it would be a good idea to explicitly compare the keys on the edited yaml fields with the original items during `apply_data()`, and abort if they differ, in order to catch this?

*  A minor presentational issue on the generated yaml: would it be desirable to use an `OrderedDict` (combined with the `Dumper` from `confit`) for the yaml items in the hopes of preserving the field order specified by the user on `config.itemfields`?

sampsyo commented:
Awesome! Thank you for the new tests!

> test_single_title_edit_apply is failing due to the assertion on the number of calls to Item.write() (it expects 1 call to write(), and recording 10).

Good question. We could add "dirty" checking to only write objects that have actually changed. That would help with performance too.

> The new test (test_invalid_yaml_extra_field, which tries to mimic the user adding random fields to the yaml) results in an error, as the plugin marks it as a valid change during EditPlugin.save_write() and tries to act accordingly:

I'm actually in favor of letting you use the plugin to add fields! It seems useful, for example, in the scenario from #1723. Does that make sense to you?

> A minor presentational issue on the generated yaml: would it be desirable to use an OrderedDict (combined with the Dumper from confit) for the yaml items in the hopes of preserving the field order specified by the user on config.itemfields?

Yeah, preserving order would be good. I'm a little hesitant to introduce a dependency on Confit's YAML dumper just now, though, since we've recently discussed replacing it with a third-party library... maybe we should hold off on this nicety for now.

diego-plan9 commented:
> I'm actually in favor of letting you use the plugin to add fields! It seems useful, for example, in the scenario from #1723. Does that make sense to you?

Sure, it does make sense! One could argue that the "canonical" way to do so might be that the user explicitly includes his desired fields on his `config.yaml` (or via the `--extra` parameter), but I see your point and I'm happy with the "*more power to the user*" approach - I'll adjust the test case to conform with your suggested behaviour.

> Yeah, preserving order would be good. I'm a little hesitant to introduce a dependency on Confit's YAML dumper just now, though, since we've recently discussed replacing it with a third-party library... maybe we should hold off on this nicety for now.

Noted - it might just be the case of me being a bit overprotective about the prospective users at this stage, holding off the change until a decision on `rueml.yaml` or a similar package sounds good!

sampsyo commented:
I made a change to the way prompts work in the latest version: the prompt after editing the file now *shows you the pending changes* and ask you to pick from "continue editing," "apply," and "cancel." The idea is that supersedes the earlier "are you done?" prompt as well as the "confirm these are the changes you wanted" into a single prompt that lets you make a decision based on more context.

@jmwatte, I haven't yet merged over the `--group` option yet -- I'd like to merge this v1.0 into mainline beets ASAP so people can give it a try, and then we can work on that extension. (And incidentally, what do you think about calling it `--batch`?)

diego-plan9 commented:
Nice changes, and a small but significant step towards improving the user experience even at this initial stage (as I also feel that the scenario experienced by @jmwatte at #1725 is quite common and frustrating). Looking forward to having the plugin included into mainline and start shaping up the roadmap and future features!

sampsyo commented:
Hi, @pkess—thanks! There was a nefarious typo in there.

Feel free to discuss here, unless there's something longer-term that we should consider after merging in v1.0 of the plugin.

pkess commented:
sadly there are two more things:
1. when i add a field in the editor like this:
onplayer: true
the plugin sets the field onplayer to value "1". With `beet modify -w onplayer:true` the field is set to true as expected
2. when i am using the "-e" commandline switch i can't use "user-defined" fields like my onplayer field. It seems like only fields from the items table in db are supported

sampsyo commented:
Thanks for trying this out!

> 2. when i am using the "-e" commandline switch i can't use "user-defined" fields like my onplayer field. It seems like only fields from the items table in db are supported

Yep, this was an oversight. I just changed the option to work for any field.

I also changed the name from `--extra` or `-e` to `--field` or `-f`, which I think is a little easier to guess.

sampsyo commented:
> 1. when i add a field in the editor like this:
> onplayer: true
> the plugin sets the field onplayer to value "1". With beet modify -w onplayer:true the field is set to true as expected

This is another good point—the plugin is just trusting the YAML type of each value, rather than using the beets logic for parsing strings (as the `modify` plugin does).

This will, unfortunately, be a little bit difficult to fix. The problem is that the YAML parser sees the values not as strings but as a whole bouquet of different types. `foo: 123` gives you an integer, `foo: true` gives you a boolean, etc. We don't get to see the raw strings.

We'll need to resort to some sort of hack to let beets do the value parsing instead of the YAML library. Or we could switch away from YAML as the serialization format.

diego-plan9 commented:
An observation on the scenario presented by @pkess, which I'm unsure if it is the expected behaviour: if the user-specified extra field (`-f nonexistingfield`) is not already present on the item, it is currently **not** dumped into the yaml (`flatten()` silently swallows it due to iterating on `obj.items()`). I'm wondering if it they key should be added to the yaml with an empty value instead?

diego-plan9 commented:
As for the problem of yaml auto-magically performing type conversion, looks like a nontrivial issue indeed! However, I'm wondering if it could be alleviated to some extent by trying to make use of Flexible fields and `dbcore.types`, by something along the lines of:

1) Allowing the user to specified custom fields and their type on their config (such as via a small plugin):
```
custom_fields_plugin:
  onplayer: boolean
```
2) Make those available as that plugin's `CustomFieldsPlugin.item_types`.
3) Rely on the magic of `dbcore.types` for casting to the right type before yaml dumping / after yaml loading.

An informal and dirty hack (by appending `self.__class__.item_types = {'onplayer': types.Boolean()}` to `EditPlugin.__init__()`) seems to result in the yaml correctly displaying `onplayer: true` instead of `onplayer: '1'`), but I'm not sure if:
* this approach is sound
* it is useful with other `type`s, and solves more problems that it might cause
* there are enough users using custom fields to justify the extra plugin

sampsyo commented:
> An observation on the scenario presented by @pkess, which I'm unsure if it is the expected behaviour: if the user-specified extra field (-f nonexistingfield) is not already present on the item, it is currently not dumped into the yaml (flatten() silently swallows it due to iterating on obj.items()). I'm wondering if the key should be added to the yaml with an empty value instead?

Yes, I was thinking the same thing. The only trick with that change is that we'll need the "round trip" to stay correct: if you leave the field empty in the editor, beets shouldn't add it to the item. It's not quite clear to me how to distinguish that from adding the field with an empty value.

sampsyo commented:
> 1) Allowing the user to specify his own custom fields and their type on his config (such as via a small plugin):

Excellent idea! :smiley: http://docs.beets.io/en/v1.3.15/plugins/types.html

> 3) Rely on the magic of dbcore.types for casting to the right type before yaml dumping / after yaml loading.

Yes, now that you mention that, it could be as easy as just dumping everything to a string and then loading it back from a string. The only problem is if someone comes along and sees:

    foo: "true"

which is a string value and changes it to:

    foo: false

in an attempt to flip the Boolean. Then YAML will parse this not as the string "false" but as the Boolean value `False`. Sadly, dbcore's type-based string parsing infrastructure can't handle this---it wants the original string.

diego-plan9 commented:
> Excellent idea! ￼ http://docs.beets.io/en/v1.3.15/plugins/types.html

Well, that's embarrassing :sweat:. I did some greping and peek at the issues to find out if that was already in place, but seems it got past my not so deep research!

Maybe we should preliminary add some comments on the documentation of the plugin, cautioning about the potential issues with custom fields?

sampsyo commented:
Here's another way to make things safe: let's dump everything to strings, as planned, and then reject modifications like the example I showed (where the user inputs a non-string type). Over time, we can make this more tolerant by accepting, for example, the Boolean `True` as a synonym for the string "True".

sampsyo commented:
OK, here's an implementation of that idea that I'm reasonably happy with. If the user messes things up and uses a non-string value, we currently just try stringifying the value and *then* parsing it. This could be a little smarter, but I think this should work for most cases (and should solve the original problem).

One trick here is that we need to be sure that all data types can "round-trip" for ergonomic editing. One problem I ran into immediately was with the bitrate field: it gets formatted as "128kbps" (for example), but the parser just sees this as the number 128 instead of 128000. This is problematic because there is actually some data lost here---the low bits of the bitrate number.

sampsyo commented:
One way to cope with the above problem would be to special-case some types that we know are safe for YAML editing, like the numeric types or booleans. These would somehow bypass the formatting when printed. If they come back as the same type as they were written out, then they also bypass parsing; if the user changes to a different type, then we can stringify and parse as a fallback.

sampsyo commented:
OK, sorry for all the noise but this latest revision should do it. Most values are stringified and parsed again, but certain "safe" types get a pass. They are still validated to ensure nothing crazy happens, though, and parsed as strings if necessary.

diego-plan9 commented:
Looking good! I'm wondering if it would be a good time to include some more tests focused on specific functions on the plugin, or perhaps it's better to hold off until it is included on the main branch?

sampsyo commented:
Awesome! More tests are always better, but I don't have a strong opinion about whether they should come before or after merge. If it's looking reasonably complete, I think we can merge at will!

pkess commented:
Hi,

i just merged master into this branch because i experienced errors with this branch and the badfiles plugin because of unicode characters

sampsyo commented:
Thanks for updating the PR, @pkess.

I think this is ready enough for inclusion in the next point release. There are still more things we could do, of course, but no need to hold it back any more. Merging now.
